src/rgw/rgw_sal.h

```diff
@@ -145,25 +210,43 @@ class RGWBucketList {
   RGWBucketList() : buckets(), truncated(false) {}
   RGWBucketList(RGWBucketList&&) = default;
   RGWBucketList& operator=(const RGWBucketList&) = default;
-  ~RGWBucketList();
+  ~RGWBucketList() { clear(); }
 
   map<string, RGWBucket*>& get_buckets() { return buckets; }
   bool is_truncated(void) const { return truncated; }
   void set_truncated(bool trunc) { truncated = trunc; }
   void add(RGWBucket* bucket) {
-    buckets[bucket->ent.bucket.name] = bucket;
+    buckets[bucket->info.bucket.name] = bucket;
   }
   size_t count() const { return buckets.size(); }
-  void clear() { buckets.clear(); truncated = false; }
-}; // class RGWBucketList
+  void clear(void) {
+    for (auto i = buckets.begin(); i != buckets.end(); i++)
+      delete i->second;
+    buckets.clear();
+    truncated = false;
+  }
+};
 
 class RGWObject {
   protected:
     rgw_obj_key key;
+    RGWBucket* bucket;
+    std::string index_hash_source;
+    uint64_t obj_size;
+    ceph::real_time mtime;
 
   public:
-    RGWObject() : key() {}
-    RGWObject(const rgw_obj_key& _k) : key(_k) {}
+    RGWObject() : key(), bucket(nullptr), index_hash_source(), obj_size(), mtime() {}
+    RGWObject(const rgw_obj_key& _k) : key(_k), bucket(), index_hash_source(), obj_size(), mtime() {}
+    RGWObject(const rgw_obj_key& _k, RGWBucket* _b) : key(_k), bucket(_b), index_hash_source(), obj_size(), mtime() {}
+    RGWObject(RGWObject& _o)
```

Comment:
copy ctors should take the source object by const ref. and if you're just doing a simple memberwise copy, you may as well use the default: `RGWObject(const RGWObject&) = default;`